examples/energy-management-app/energy-management-common/src/EnergyEvseDelegateImpl.cpp

```diff
@@ -831,17 +1420,142 @@ DataModel::Nullable<CharSpan> EnergyEvseDelegate::GetVehicleID()
 /* Session SESS attributes */
 DataModel::Nullable<uint32_t> EnergyEvseDelegate::GetSessionID()
 {
-    return mSessionID;
+    return mSession.mSessionID;
 }
 DataModel::Nullable<uint32_t> EnergyEvseDelegate::GetSessionDuration()
 {
-    return mSessionDuration;
+    mSession.RecalculateSessionDuration();
+    return mSession.mSessionDuration;
 }
 DataModel::Nullable<int64_t> EnergyEvseDelegate::GetSessionEnergyCharged()
 {
-    return mSessionEnergyCharged;
+    return mSession.mSessionEnergyCharged;
 }
 DataModel::Nullable<int64_t> EnergyEvseDelegate::GetSessionEnergyDischarged()
 {
-    return mSessionEnergyDischarged;
+    return mSession.mSessionEnergyDischarged;
+}
+
+/**
+ * @brief   Helper function to get current timestamp in Epoch format
+ *
+ * @param   chipEpoch reference to hold return timestamp
+ */
+CHIP_ERROR GetEpochTS(uint32_t & chipEpoch)
+{
+    chipEpoch = 0;
+
+    System::Clock::Milliseconds64 cTMs;
+    CHIP_ERROR err = System::SystemClock().GetClock_RealTimeMS(cTMs);
+
+    /* If the GetClock_RealTimeMS returns CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE, then
+     * This platform cannot ever report real time !
+     * This should not be certifiable since getting time is a Mandatory
+     * feature of EVSE Cluster
+     */
+    VerifyOrDie(err != CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE);
+
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Zcl, "EVSE: Unable to get current time - err:%" CHIP_ERROR_FORMAT, err.Format());
+        return err;
+    }
+
+    auto unixEpoch = std::chrono::duration_cast<System::Clock::Seconds32>(cTMs).count();
+    if (!UnixEpochToChipEpochTime(unixEpoch, chipEpoch))
+    {
+        ChipLogError(Zcl, "EVSE: unable to convert Unix Epoch time to Matter Epoch Time");
+        return err;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+/**
+ * @brief This function samples the start-time, and energy meter to hold the session info
+ *
+ * @param chargingMeterValue    - The current value of the energy meter (charging) in mWh
+ * @param dischargingMeterValue - The current value of the energy meter (discharging) in mWh
+ */
+void EvseSession::StartSession(int64_t chargingMeterValue, int64_t dischargingMeterValue)
+{
+    /* Get Timestamp */
+    uint32_t chipEpoch = 0;
+    if (GetEpochTS(chipEpoch) != CHIP_NO_ERROR)
+    {
+        ChipLogError(AppServer, "EVSE: Failed to get timestamp when starting session");
```

Comment:
There is a couple of ChipLogErrors in GetEpochTS()
See Line 1460 and 1467?

```CHIP_ERROR GetEpochTS(uint32_t & chipEpoch)
{
    chipEpoch = 0;

    System::Clock::Milliseconds64 cTMs;
    CHIP_ERROR err = System::SystemClock().GetClock_RealTimeMS(cTMs);

    /* If the GetClock_RealTimeMS returns CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE, then
     * This platform cannot ever report real time !
     * This should not be certifiable since getting time is a Mandatory
     * feature of EVSE Cluster
     */
    VerifyOrDie(err != CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE);

    if (err != CHIP_NO_ERROR)
    {
        ChipLogError(Zcl, "EVSE: Unable to get current time - err:%" CHIP_ERROR_FORMAT, err.Format());
        return err;
    }

    auto unixEpoch = std::chrono::duration_cast<System::Clock::Seconds32>(cTMs).count();
    if (!UnixEpochToChipEpochTime(unixEpoch, chipEpoch))
    {
        ChipLogError(Zcl, "EVSE: unable to convert Unix Epoch time to Matter Epoch Time");
        return err;
    }

    return CHIP_NO_ERROR;
}
```
Or did you mean somewhere else?

Reply:
Did not see that they were logged below. My general expectation is that anything that receives errors has 2 choices:
   - report then up the stack
   - drop them to the floor, but in this case log them
 
I saw your code did the drop to the floor but no logging. I am worried that if you rely that lower levels always log before returning any error then this will not be very maintainable. I would actually consider logging twice in this case. 

You could also document that `GetEpochTS(chipEpoch)`  will *always* log any returned errors so that callers do not need to log, however that seems slightly awkward to me.

Reply:
maybe add a comment that reason is logged by lower levels. at the same time I believe just double-logging the error code is probably ok too ... won't cause harm and better to be sure this can be debugged.

Reply:
Added improved error logging in both places and improved comments.